### ISSUE TITLE
Full offline mode (createSubscribeQuery)

### ISSUE THREAD

**Issue by zzph** [share/sharedb#575]
Hi,

I’m trying to create a fully offline experience, not only storing the ops to send- but reading data from local storage (indexdb) when there’s no connection to the server. 

Is this possible? Do you have a suggestion on implantation?

My attempt would bf something like:

1. for every item, to store the following in local storage:

{
          data,
          src: connection.id,
          v: version,
          seq: connection.seq,
          c: collection,
          d: id,
}

2. When unable to connect to the server, read items from local storage, recreating each one as a “doc” with normal properties (submitOp etc)

**Comment by alecgibson** [comment]
If you want full offline support, including queries (which are usually handled by the DB, so not very offline-friendly!), then I have wondered in the past about running an instance of the ShareDB server in the browser (since it is, after all, JavaScript). It would have query support thanks to [`sharedb-mingo-memory`](https://github.com/share/sharedb-mingo-memory/).

The tricky part is that you'd still need to figure out how to sync your in-memory DB state with indexdb, which isn't a solved problem as far as ShareDB goes, and I sadly don't have any time to work on that personally.

**Comment by zzph** [comment]
Hmm, interested. I'm keen to spend time figuring it out.

Could elaborate how I'd "mock" the results coming through on connection from indexDB (as opposed to remotely)? Where is a good starting point?

**Comment by alecgibson** [comment]
You could probably start by just copy-pasting and/or forking [`sharedb-mingo-memory`](https://github.com/share/sharedb-mingo-memory/blob/master/index.js). You can then extend its functionality to actively write to indexdb every time it writes to memory (and to fetch from indexdb on initialisation, or if something isn't already loaded into memory).

You'd register it on your "backend" (which is actually running on the server) like any other [DB adapter](https://share.github.io/sharedb/adapters/database).

I forgot to mention that if you get this working fully offline, there's still some work to do once you come back online to reconcile your local database with the real database, since your document histories will have diverged. I haven't thought too hard on that yet, but I guess we can cross that bridge once you get your local DB running properly 😅

**Comment by zzph** [comment]
Thanks- I'm  a little confused how I'd run it on the 'backend' though- this would have to be client side (browser) offline mode.

Did you mean to run the server code  in my browser? If so, how do I get the "socket" connected to it?

```
  const backend = new ShareDB({
    db: new MyIndexDBFork(),
  })
  const connection = backend.connect()
  
  // 
  ```

**Comment by alecgibson** [comment]
> Did you mean to run the server code in my browser

Yes, this is exactly what I mean. If you run the server code in your browser, out-of-the-box you can already call `backend.connect()` with no arguments and it will create an in-memory [`StreamSocket`](https://github.com/share/sharedb/blob/8b531bedf19860ffe0b37a3e1c8da7a32b5005bd/lib/backend.js#L126) and [return](https://github.com/share/sharedb/blob/8b531bedf19860ffe0b37a3e1c8da7a32b5005bd/lib/backend.js#L145) a `Connection` instance.

**Comment by zzph** [comment]
I've had a go, firstly needing to `npm install stream` as it couldn't be found. 

However, I get the error `TypeError: Cannot read properties of undefined (reading 'call') at new ServerStream`. Any ideas why?

```
  const backend = new ShareDB({db: new MyIndexDBFork() })
  const connection = backend.connect()
  connection.createSubscribeQuery
```
![Screen Shot 2022-09-19 at 10 45 20 pm Large](https://user-images.githubusercontent.com/84111615/191001175-2467bb8c-82e1-4562-b96a-3882884e8643.jpeg)

**Comment by alecgibson** [comment]
I don't think you want `stream` (it's last update was 8yrs ago); we're using [`stream-browserify`](https://www.npmjs.com/package/stream-browserify)

**Comment by zzph** [comment]
> I don't think you want `stream` (it's last update was 8yrs ago); we're using [`stream-browserify`](https://www.npmjs.com/package/stream-browserify)

I tried installing `stream-browsrify` but still got the same error.

Hmm seems it's still using stream, no?

Eg here:
https://github.com/share/sharedb/blob/e29936ca2dffd187b3fcfa3ca1b7e83becb6a2dc/lib/op-stream.js#L2

**Comment by alecgibson** [comment]
Yes, you'll need to include this in whatever you're using to package your JS. For example, if using Webpack 5: https://webpack.js.org/configuration/resolve/#resolvefallback

**Comment by zzph** [comment]
Hmm I'm using plain js (no bundler), i can't figure it out I thought exports would do it in package.json

**Comment by alecgibson** [comment]
Unfortunately you currently need some sort of bundler for ShareDB. See https://github.com/share/sharedb/issues/255

**Comment by zzph** [comment]
As an alternative,  can you offer any suggestion on the below example? It work well, except the `submitOps` method inside the class isn't actually updating the data. Maybe I dont understand JS classes correctly. Any ideas?

> export class DocMock {
>   constructor(collection, data) {
>     this.connection = { canSend: false }
>     this.collection = collection
>     this.data = data
>     this.id = data.id
>   }
> 
>   submitOp(ops, cb) {
>     this.data = getItemWithOps(this.data, ops)
>     this.emit('op', ops, true)
>     cb(null)
>   }
> 
>   on(eventType, cb) {
>     this[`_on${eventType}`] = cb;
>   }
> 
>   emit(eventType, a, b) {
>     if (this[`_on${eventType}`]) {
>       this[`_on${eventType}`](a, b);
>     }
>   }
